### PR TITLE
feat(unstable): add metrics to otel

### PR DESCRIPTION
**Pull request by devsnek**
Refs: https://github.com/denoland/deno/issues/26852

Initial support for exporting metrics.

### CODE REVIEW

File: ext/telemetry/lib.rs [hunk 1]
```diff
@@ -322,8 +343,69 @@ mod hyper_client {
...
+  Flush(tokio::sync::oneshot::Sender<()>),
+}
+
+struct MetricProcessor {
+  tx: tokio::sync::mpsc::Sender<MetricProcessorMessage>,
+}
+
+impl MetricProcessor {
+  fn new(exporter: MetricExporter) -> Self {
+    let (tx, mut rx) = tokio::sync::mpsc::channel(2048);
```
Comment (bartlomieju) [inline ext/telemetry/lib.rs:357]: Is there any reasoning why buffer of this size was chosen?

File: ext/telemetry/lib.rs [hunk 2]
```diff
@@ -322,8 +343,69 @@ mod hyper_client {
...
+  fn submit(&self, rm: ResourceMetrics) {
+    let _ = self
+      .tx
+      .try_send(MetricProcessorMessage::ResourceMetrics(rm));
+  }
+
+  fn force_flush(&self) -> Result<(), anyhow::Error> {
+    let (tx, rx) = tokio::sync::oneshot::channel();
+    self.tx.try_send(MetricProcessorMessage::Flush(tx))?;
+    deno_core::futures::executor::block_on(rx)?;
```
Comment (bartlomieju) [inline ext/telemetry/lib.rs:397]: Using `block_on` is usually trouble, but I don't see any other way here...

File: ext/telemetry/lib.rs [hunk 3]
```diff
@@ -678,12 +776,25 @@ fn op_otel_log(
...
+  scope: &mut v8::HandleScope<'s>,
+  string: v8::Local<'s, v8::String>,
+) -> String {
+  let x = v8::ValueView::new(scope, string);
+  match x.data() {
+    v8::ValueViewData::OneByte(bytes) => {
+      String::from_utf8_lossy(bytes).into_owned()
+    }
+    v8::ValueViewData::TwoByte(bytes) => String::from_utf16_lossy(bytes),
+  }
```
Comment (bartlomieju) [inline ext/telemetry/lib.rs:795]: Is there a reason to do it instead of `v8::String::to_rust_string_lossy()`?

File: ext/telemetry/lib.rs [hunk 4]
```diff
@@ -866,9 +969,596 @@ fn op_otel_span_flush(state: &mut OpState) {
...
 
-  let Some((span_processor, _)) = OTEL_PROCESSORS.get() else {
+  let Some(Processors { spans, .. }) = OTEL_PROCESSORS.get() else {
+    return;
+  };
+
+  spans.on_end(temporary_span.0);
+}
+
+pub struct TemporaryMetricsExport {
```
Comment (bartlomieju) [inline ext/telemetry/lib.rs:979]: Could you add some docstrings here about the purpose of these structs and enums?

File: ext/telemetry/lib.rs [hunk 5]
```diff
@@ -866,9 +969,596 @@ fn op_otel_span_flush(state: &mut OpState) {
...
+  };
+
+  if let Some(TemporaryMetric {
+    data: TemporaryMetricData::Histogram(histogram),
+    ..
+  }) = &mut temp.metric
+  {
+    histogram
+      .data_points
+      .last_mut()
```
Comment (bartlomieju) [inline ext/telemetry/lib.rs:1363]: Is it guaranteed there's gonna be at least one element?

File: ext/telemetry/lib.rs (hunk 3 above)
Comment (devsnek) [reply ext/telemetry/lib.rs:795]: https://github.com/denoland/rusty_v8/pull/1645

File: ext/telemetry/lib.rs (hunk 1 above)
Comment (devsnek) [reply ext/telemetry/lib.rs:357]: its the default from the other two processors

File: ext/telemetry/lib.rs (hunk 2 above)
Comment (devsnek) [reply ext/telemetry/lib.rs:397]: the other two processors use block_on as well

File: ext/telemetry/lib.rs (hunk 5 above)
Comment (devsnek) [reply ext/telemetry/lib.rs:1363]: `op_otel_metrics_histogram_data_point` is always called before this

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**